### PR TITLE
Further settings for the map element list and more map elements by adjustable hit range

### PR DESCRIPTION
**Pull request by ThHanika**
The trigger for this expansion was the use of openvario without touch and without mouse operation. It was very difficult to use the center crosshair on the screen to click on a waypoint or airfield to display the map item list. Even a subsequently installed resistive touchscreen has not brought benefits because of its inaccuracy.

With this extension, you can now expand the area around the crosshairs, even so far that all visible elements can be listed by clicking on the map. Even during the flight in turbulent air, this extension has proven itself.

The maximum number of the list is 32 items and is sorted by distance.

### CODE REVIEW

File: src/MapWindow/GlueMapWindowItems.cpp [hunk 1]
```diff
@@ -49,6 +49,10 @@ GlueMapWindow::ShowMapItems(const GeoPoint &location,
   const MoreData &basic = CommonInterface::Basic();
   const DerivedInfo &calculated = CommonInterface::Calculated();
 
+  bool doRefreshMapList = false;
+
+  do{
+
   bool only_landable = false;
```
Comment (MaxKellermann) [inline src/MapWindow/GlueMapWindowItems.cpp]: Bad indent

File: src/MapWindow/GlueMapWindowItems.cpp [hunk 2]
```diff
@@ -49,6 +49,10 @@ GlueMapWindow::ShowMapItems(const GeoPoint &location,
   const MoreData &basic = CommonInterface::Basic();
   const DerivedInfo &calculated = CommonInterface::Calculated();
 
+  bool doRefreshMapList = false;
+
+  do{
```
Comment (MaxKellermann) [inline src/MapWindow/GlueMapWindowItems.cpp]: Bad coding style

File: src/MapWindow/GlueMapWindowItems.cpp [hunk 3]
```diff
@@ -49,6 +49,10 @@ GlueMapWindow::ShowMapItems(const GeoPoint &location,
   const MoreData &basic = CommonInterface::Basic();
   const DerivedInfo &calculated = CommonInterface::Calculated();
 
+  bool doRefreshMapList = false;
```
Comment (MaxKellermann) [inline src/MapWindow/GlueMapWindowItems.cpp]: No camelCase for variable names

File: src/Dialogs/MapItemListDialog.hpp [hunk 4]
```diff
@@ -32,7 +32,7 @@ struct FinalGlideBarLook;
 struct MapSettings;
 class ProtectedAirspaceWarningManager;
 
-void
+bool
```
Comment (MaxKellermann) [inline src/Dialogs/MapItemListDialog.hpp]: Documentation for return value missing

File: src/MapWindow/GlueMapWindowItems.cpp [hunk 5]
```diff
@@ -131,10 +135,12 @@ GlueMapWindow::ShowMapItems(const GeoPoint &location,
     return false;
   }
 
-  ShowMapItemListDialog(list,
+  doRefreshMapList = ShowMapItemListDialog(list,
                         UIGlobals::GetDialogLook(), look, traffic_look,
                         final_glide_bar_renderer.GetLook(), settings,
                         glide_computer != nullptr
                         ? &glide_computer->GetAirspaceWarnings() : nullptr);
+  }while(doRefreshMapList);
```
Comment (MaxKellermann) [inline src/MapWindow/GlueMapWindowItems.cpp]: Bad coding style

File: src/Dialogs/MapItemListDialog.cpp [hunk 6]
```diff
@@ -345,20 +351,27 @@ ShowMapItemListDialog(const MapItemList &list,
...
                                   traffic_look, final_glide_look, settings);
-    assert(i >= -1 && i < (int)list.size());
+
+    assert(i >= -2 && i < (int)list.size());
+
     if (i >= 0)
       ShowMapItemDialog(*list[i], airspace_warnings);
+
+    /* setting of map item list was done */
+    if(i == -2) return true;
```
Comment (MaxKellermann) [inline src/Dialogs/MapItemListDialog.cpp]: What is this `-2`?

File: src/Dialogs/MapItemListDialog.cpp [hunk 7]
```diff
@@ -333,7 +338,8 @@ ShowMapItemDialog(const MapItem &item,
   }
 }
 
-void
+/* return value signals the using of map item list settings => refresh map item list */
```
Comment (MaxKellermann) [inline src/Dialogs/MapItemListDialog.cpp]: Bad style for API documentation, and belongs to prototype.

File: src/Dialogs/MapItemListDialog.cpp [hunk 8]
```diff
@@ -96,6 +96,7 @@ class MapItemListWidget final
 
 public:
   void CreateButtons(WidgetDialog &dialog);
+  bool bSettingDone = false; // signals the setting of MapItemList paramter
```
Comment (MaxKellermann) [inline src/Dialogs/MapItemListDialog.cpp]: No camelCase, bad API documentation style.

File: src/Dialogs/MapItemListDialog.cpp [hunk 9]
```diff
@@ -251,6 +252,7 @@ MapItemListWidget::OnAction(int id)
   switch (id) {
   case SETTINGS:
     ShowMapItemListSettingsDialog();
+    Profile::Save();
```
Comment (MaxKellermann) [inline src/Dialogs/MapItemListDialog.cpp]: Why?

File: src/Dialogs/MapItemListSettingsPanel.cpp [hunk 10]
```diff
@@ -37,15 +37,36 @@ MapItemListSettingsPanel::Prepare(ContainerWindow &parent, const PixelRect &rc)
 
   const MapSettings &settings = CommonInterface::GetMapSettings();
 
-  AddBoolean(_("Show Location row"),
+  AddBoolean(_("Lists Location row"),
```
Comment (MaxKellermann) [inline src/Dialogs/MapItemListSettingsPanel.cpp]: Why?

File: src/Dialogs/MapItemListSettingsPanel.cpp [hunk 11]
```diff
@@ -37,15 +37,36 @@ MapItemListSettingsPanel::Prepare(ContainerWindow &parent, const PixelRect &rc)
...
              _("If enabled a row at the top will be added showing you the "
                "distance and bearing to the location and the elevation."),
              settings.item_list.add_location);
 
-  AddBoolean(_("Show Arrival Altitude"),
+  AddBoolean(_("Lists Arrival Altitude"),
```
Comment (MaxKellermann) [inline src/Dialogs/MapItemListSettingsPanel.cpp]: Why?

File: src/MapSettings.hpp [hunk 12]
```diff
@@ -78,6 +78,18 @@ struct MapItemListSettings {
...
   bool add_arrival_altitude;
 
+  /** Add an Airspace to the MapItemList? */
+  bool add_airspace;
+
+  /** Add flarm-traffic to the MapItemList? */
+  bool add_traffic;
+
+  /* nearest map items: the range is a circle calculate in % of the smallest side of the display */
+  int range_of_nearest_map_items_in_percent_of_displaysize;
```
Comment (MaxKellermann) [inline src/MapSettings.hpp]: Why is this signed?

File: src/MapSettings.hpp [hunk 13]
```diff
@@ -78,6 +78,18 @@ struct MapItemListSettings {
...
+
+  /* nearest map items: range-filter in km for only landable waypoints  0km only landable */
+  int rangefilter_all_waypoint_up_to_km;
```
Comment (MaxKellermann) [inline src/MapSettings.hpp]: Why is this signed?

File: src/MapSettings.cpp [hunk 14]
```diff
@@ -28,6 +28,10 @@ MapItemListSettings::SetDefaults()
 {
   add_location = true;
   add_arrival_altitude = true;
+  add_airspace = true;
+  add_traffic = true;
+  range_of_nearest_map_items_in_percent_of_displaysize = 0; // 0% default behavior
+  rangefilter_all_waypoint_up_to_km = 100; // list all waypoints
```
Comment (MaxKellermann) [inline src/MapSettings.cpp]: Why does "100km" list all waypoints?

File: src/MapWindow/GlueMapWindowItems.cpp [hunk 15]
```diff
@@ -67,27 +76,31 @@ GlueMapWindow::ShowMapItems(const GeoPoint &location,
...
   if (waypoints)
-    builder.AddWaypoints(*waypoints);
+    builder.AddWaypoints(*waypoints,only_landable);
 
 #ifdef HAVE_NOAA
   if (noaa_store)
     builder.AddWeatherStations(*noaa_store);
 #endif
 
-  builder.AddTraffic(basic.flarm.traffic);
```
Comment (MaxKellermann) [inline src/MapWindow/GlueMapWindowItems.cpp]: This changes the order of items. Why?

File: src/MapWindow/GlueMapWindowItems.cpp [hunk 16]
```diff
@@ -67,27 +76,31 @@ GlueMapWindow::ShowMapItems(const GeoPoint &location,
...
     builder.AddTaskOZs(*task);
 
-  const Airspaces *airspace_database = airspace_renderer.GetAirspaces();
-  if (airspace_database)
-    builder.AddVisibleAirspace(*airspace_database,
-                               airspace_renderer.GetWarningManager(),
-                               computer_settings.airspace,
-                               settings.airspace, basic,
-                               calculated);
+  if(settings.item_list.add_airspace) {
```
Comment (MaxKellermann) [inline src/MapWindow/GlueMapWindowItems.cpp]: Bad coding style

File: src/Dialogs/MapItemListSettingsPanel.cpp [hunk 17]
```diff
@@ -37,15 +37,36 @@ MapItemListSettingsPanel::Prepare(ContainerWindow &parent, const PixelRect &rc)
...
+             _("Flarm-traffic items are listed."),
+             settings.item_list.add_traffic);
+
+  AddInteger(_("Hit Range on the Screen"),
+             _("Lists more map items with a radius in% of screen width around the center of the screen. "
+                 "Useful for pure key control. 0% defines the default behavior, "
+                 "at 100% a radius is taken into account until the next screen edge. "
+                 "125% is to be used to list all displayed map items on the screen. "
+                 "A maximum radius of 50km radius is calculated. A mouse click determines the center point."),_("%d %%"),_("%d %%"),0,125,25,
+               settings.item_list.range_of_nearest_map_items_in_percent_of_displaysize );
```
Comment (MaxKellermann) [inline src/Dialogs/MapItemListSettingsPanel.cpp]: Bad coding style

File: src/Dialogs/MapItemListSettingsPanel.cpp [hunk 18]
```diff
@@ -37,15 +37,36 @@ MapItemListSettingsPanel::Prepare(ContainerWindow &parent, const PixelRect &rc)
...
+
+  AddInteger(_("Only Landable on Hit Range>"),
+             _("As the zoom level decreases, the calculated range of hits increases in kilometers. "
+                 "From this limit only landable will be listed."),_("%d km"),_("%d km"),0,50,10,
+               settings.item_list.rangefilter_all_waypoint_up_to_km );
```
Comment (MaxKellermann) [inline src/Dialogs/MapItemListSettingsPanel.cpp]: Bad coding style

File: src/Dialogs/MapItemListSettingsPanel.cpp [hunk 19]
```diff
@@ -37,15 +37,36 @@ MapItemListSettingsPanel::Prepare(ContainerWindow &parent, const PixelRect &rc)
...
+             _("Airspace are listed"),
+             settings.item_list.add_airspace);
+
+  AddBoolean(_("Lists Traffic"),
+             _("Flarm-traffic items are listed."),
+             settings.item_list.add_traffic);
+
+  AddInteger(_("Hit Range on the Screen"),
+             _("Lists more map items with a radius in% of screen width around the center of the screen. "
+                 "Useful for pure key control. 0% defines the default behavior, "
```
Comment (MaxKellermann) [inline src/Dialogs/MapItemListSettingsPanel.cpp]: Wrong indent

File: src/Dialogs/MapItemListSettingsPanel.cpp [hunk 20]
```diff
@@ -37,15 +37,36 @@ MapItemListSettingsPanel::Prepare(ContainerWindow &parent, const PixelRect &rc)
...
+                 "at 100% a radius is taken into account until the next screen edge. "
+                 "125% is to be used to list all displayed map items on the screen. "
+                 "A maximum radius of 50km radius is calculated. A mouse click determines the center point."),_("%d %%"),_("%d %%"),0,125,25,
+               settings.item_list.range_of_nearest_map_items_in_percent_of_displaysize );
+
+  AddInteger(_("Only Landable on Hit Range>"),
+             _("As the zoom level decreases, the calculated range of hits increases in kilometers. "
+                 "From this limit only landable will be listed."),_("%d km"),_("%d km"),0,50,10,
```
Comment (MaxKellermann) [inline src/Dialogs/MapItemListSettingsPanel.cpp]: Wrong indent

File: src/MapWindow/GlueMapWindowItems.cpp [hunk 21]
```diff
@@ -49,7 +49,16 @@ GlueMapWindow::ShowMapItems(const GeoPoint &location,
   const MoreData &basic = CommonInterface::Basic();
   const DerivedInfo &calculated = CommonInterface::Calculated();
 
-  auto range = visible_projection.DistancePixelsToMeters(Layout::GetHitRadius());
+  bool only_landable = false;
+  int rangedisplaypercent = settings.item_list.range_of_nearest_map_items_in_percent_of_displaysize;
+  auto range = 0.0;
+  if(rangedisplaypercent == 0) range = visible_projection.DistancePixelsToMeters(Layout::GetHitRadius());
```
Comment (MaxKellermann) [inline src/MapWindow/GlueMapWindowItems.cpp]: Bad coding style, and what does this magic value 0 mean? This looks wrong.

File: src/MapWindow/GlueMapWindowItems.cpp [hunk 22]
```diff
@@ -49,7 +49,16 @@ GlueMapWindow::ShowMapItems(const GeoPoint &location,
...
+  else {
+    range = visible_projection.DistancePixelsToMeters(Layout::min_screen_pixels/2*rangedisplaypercent/100);
```
Comment (MaxKellermann) [inline src/MapWindow/GlueMapWindowItems.cpp]: Should use SI units. The `range` variable appears to be in kilometres, and that is just asking for bugs.

File: src/MapWindow/GlueMapWindowItems.cpp [hunk 23]
```diff
@@ -107,7 +120,7 @@ GlueMapWindow::ShowMapItems(const GeoPoint &location,
   }
 
   // Sort the list of map items
-  list.Sort();
+  // Sortierung aufheben, es wird nach entfernung sortiert list.Sort();
```
Comment (MaxKellermann) [inline src/MapWindow/GlueMapWindowItems.cpp]: No German!

File: src/MapWindow/Items/Builder.cpp [hunk 24]
```diff
@@ -40,13 +40,46 @@ class WaypointListBuilderVisitor:
   public WaypointVisitor
 {
   MapItemList &list;
+  bool only_landable;
+  GeoPoint location;
```
Comment (MaxKellermann) [inline src/MapWindow/Items/Builder.cpp]: Documentation?

File: src/MapWindow/Items/Builder.cpp [hunk 25]
```diff
@@ -40,13 +40,46 @@ class WaypointListBuilderVisitor:
...
+
+  // distance waypoint vector, for sorting on distance in destructor
+  std::vector<std::pair<double,std::shared_ptr<const Waypoint>> > tempMIvect;
```
Comment (MaxKellermann) [inline src/MapWindow/Items/Builder.cpp]: Bad variable name

File: src/MapWindow/Items/Builder.cpp [hunk 26]
```diff
@@ -40,13 +40,46 @@ class WaypointListBuilderVisitor:
   public WaypointVisitor
 {
   MapItemList &list;
+  bool only_landable;
+  GeoPoint location;
+
+  // distance waypoint vector, for sorting on distance in destructor
```
Comment (MaxKellermann) [inline src/MapWindow/Items/Builder.cpp]: Bad documentation style

File: src/MapWindow/Items/Builder.cpp [hunk 27]
```diff
@@ -40,13 +40,46 @@ class WaypointListBuilderVisitor:
...
+  std::vector<std::pair<double,std::shared_ptr<const Waypoint>> > tempMIvect;
 
 public:
-  WaypointListBuilderVisitor(MapItemList &_list):list(_list) {}
+  WaypointListBuilderVisitor(MapItemList &_list, bool _only_landable = false, GeoPoint _location = GeoPoint().Invalid() ):list(_list),only_landable(_only_landable),location(_location) {}
```
Comment (MaxKellermann) [inline src/MapWindow/Items/Builder.cpp]: Line too long

File: src/MapWindow/Items/Builder.cpp [hunk 28]
```diff
@@ -40,13 +40,46 @@ class WaypointListBuilderVisitor:
...
+
+  void Visit(const WaypointPtr & waypointptr) override {
+    if(location.IsValid()){
```
Comment (MaxKellermann) [inline src/MapWindow/Items/Builder.cpp]: Bad coding style

File: src/MapWindow/Items/Builder.cpp [hunk 29]
```diff
@@ -40,13 +40,46 @@ class WaypointListBuilderVisitor:
...
+      double distance = waypointptr->location.Distance( location ); // calculate distance
+      std::pair<double,std::shared_ptr<const Waypoint>> elemptr = std::make_pair(distance,waypointptr); // save pair distance and waypoint
+      tempMIvect.push_back(elemptr);
```
Comment (MaxKellermann) [inline src/MapWindow/Items/Builder.cpp]: use `emplace_back()`

File: src/MapWindow/Items/Builder.cpp [hunk 30]
```diff
@@ -40,13 +40,46 @@ class WaypointListBuilderVisitor:
...
+          list.append(new WaypointMapItem(waypointptr));
+        }
+      }
+    }
+  }
 
-  void Visit(const WaypointPtr &waypoint) override {
-    if (!list.full())
-      list.append(new WaypointMapItem(waypoint));
+  virtual ~WaypointListBuilderVisitor() {
```
Comment (MaxKellermann) [inline src/MapWindow/Items/Builder.cpp]: Why virtual?

File: src/MapWindow/Items/Builder.cpp [hunk 31]
```diff
@@ -40,13 +40,46 @@ class WaypointListBuilderVisitor:
...
+    if(location.IsValid()){ // sort on distance available
+      std::sort(tempMIvect.begin(),tempMIvect.end());
```
Comment (MaxKellermann) [inline src/MapWindow/Items/Builder.cpp]: Bad coding style

File: src/MapWindow/Items/Builder.cpp [hunk 32]
```diff
@@ -40,13 +40,46 @@ class WaypointListBuilderVisitor:
...
+        std::shared_ptr<const Waypoint> WPptr = tempMIvect[i].second;
+        if (!list.full()){
+          if(only_landable){
+            if((WPptr->IsLandable()) || (tempMIvect[i].first < 250.0)) list.append(new WaypointMapItem(WPptr)); // list also this item, it was clicked!!!
+          }else{
+            list.append(new WaypointMapItem(WPptr));
+          }
+        } else break;
+      }
+      tempMIvect.clear();
```
Comment (MaxKellermann) [inline src/MapWindow/Items/Builder.cpp]: Why?

File: src/MapWindow/GlueMapWindowItems.cpp (hunk 1 above)
Comment (ThHanika) [reply src/MapWindow/GlueMapWindowItems.cpp]: Oh yes, but the correct indent would have deleted a lot of code lines and added new ones without the functional changes. I change that.

File: src/Dialogs/MapItemListDialog.hpp (hunk 4 above)
Comment (ThHanika) [reply src/Dialogs/MapItemListDialog.hpp]: OK

File: src/MapWindow/GlueMapWindowItems.cpp (hunk 5 above)
Comment (ThHanika) [reply src/MapWindow/GlueMapWindowItems.cpp]: Please, I want to learn. What is better

Comment (MaxKellermann) [reply src/MapWindow/GlueMapWindowItems.cpp]: Two spaces missing.  `} while (doRefreshMapList);`
Read the developer manual's section "Code style".

File: src/Dialogs/MapItemListDialog.cpp (hunk 6 above)
Comment (ThHanika) [reply src/Dialogs/MapItemListDialog.cpp]: I'm thinking about setting up some constants for it. The return is usually an index of the mapitemlist and -1 if the index does not exist, -2 is now the return for ShowMapItemDialog if a new settings for the list has been made.

File: src/Dialogs/MapItemListDialog.cpp (hunk 9 above)
Comment (ThHanika) [reply src/Dialogs/MapItemListDialog.cpp]: the settings are elements of the profile.

File: src/Dialogs/MapItemListSettingsPanel.cpp (hunk 10 above)
Comment (ThHanika) [reply src/Dialogs/MapItemListSettingsPanel.cpp]: Hm, the location row member of a list and in the next pull request, you can use it like a waypoint.

File: src/MapSettings.hpp (hunk 12 above)
Comment (ThHanika) [reply src/MapSettings.hpp]: OK

File: src/MapSettings.cpp (hunk 14 above)
Comment (ThHanika) [reply src/MapSettings.cpp]: I change the comment.

File: src/MapWindow/GlueMapWindowItems.cpp (hunk 15 above)
Comment (ThHanika) [reply src/MapWindow/GlueMapWindowItems.cpp]: it is moved to then beginning, flarm-traffic could be more interested during flights, Only 32 mapitems are possible, if you use a great range, there is no change for traffic-items, if waypoints filled the list.
if no interest or to much traffic, you can disable traffic-items

File: src/Dialogs/MapItemListSettingsPanel.cpp (hunk 10 above)
Comment (MaxKellermann) [reply src/Dialogs/MapItemListSettingsPanel.cpp]: Sentence no verb.

**Comment by MaxKellermann** [comment]
Your branch now contains a merge commit. Don't merge, rebase on master instead!
The commit message "changes on last review" is bad. Did you mean to amend the previous commit? This is confusing and impossible to review.

**Comment by ThHanika** [comment]
How can I correct this? I'm a git newbie....sorry

**Comment by Exadios** [comment]
So, there is an on line book that may help you. See https://git-scm.com/book/en/v2

The sections most useful in this situation will be [Reset Demystified](https://git-scm.com/book/en/v2/Git-Tools-Reset-Demystified) and [Rewriting History](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History).

**Comment by ThHanika** [comment]
OK I hope, it is OK now...

**Comment by MaxKellermann** [comment]
No, it's not rebased.

**Comment by ThHanika** [comment]
I hope the merge is gone. 

Rebasing?!? OK, than I have no chance to test my pull request... 

look at:
https://github.com/XCSoar/XCSoar/issues/157

**Comment by ThHanika** [comment]
Rebased ? !

**Comment by ThHanika** [comment]
How is it going to continue here? Is something missing?

**Comment by glidingFisi** [comment]
I tried your feature, the idea is cool and brings a really benefit, not only for openVario, also for touch screen devices.
Your implemenation worked as far I worked with it.

Here some hints for the rebase issue:
Currently there are two merge commits in the history which must be removed:
- sync with xcsoar/xcsoar Merge branch 'NearestMapItem' of https://gith…  …
- Sync with xcsoar/xcsoar

There are 2 approaches:
- Without proper Git knowledge:
Create a new branch from master, bring your changes in some proper incremential commit groups and create a new PR.
- With git:
Go back your 11 commits:
git log (count the commits you did)
git reset HEAD~11 -> 11 = count of your commits (11)
commit you changes in some proper incremential commit groups. Now execute a force push.
*Attention you must be carefully with the force push.

**Comment by ThHanika** [comment]
I do not know if the extensions meet with interest at all, so far I have tried concerning changing wishes everything. I can also understand that one should divide the patch into different functional commits, only this patch was used without git and as a patch for openvario. If there is a serious interest, then I can try again, but at the moment I still have an OV project running, which must be done before the season opener. And maybe there is one or the other, who wants to try it, then I am open for improvement.

**Comment by MaxKellermann** [comment]
Given how unclean up this whole PR is, I think it's better to open a new one once you have time to fix all the issues.

**Pull request closed without merging**